tests/meson.build

```diff
@@ -318,17 +318,89 @@ if enable_pytest
     'test_usb.py',
     'test_wallpaper.py',
   ]
+
+  template_files = [
+    'templates/access.py',
+    'templates/account.py',
+    'templates/appchooser.py',
+    'templates/background.py',
+    'templates/clipboard.py',
+    'templates/email.py',
+    'templates/filechooser.py',
+    'templates/geoclue2.py',
+    'templates/globalshortcuts.py',
+    'templates/inhibit.py',
+    'templates/__init__.py',
+    'templates/inputcapture.py',
+    'templates/lockdown.py',
+    'templates/notification.py',
+    'templates/print.py',
+    'templates/remotedesktop.py',
+    'templates/screenshot.py',
+    'templates/settings.py',
+    'templates/usb.py',
+    'templates/wallpaper.py',
+  ]
+
   foreach pytest_file : pytest_files
     testname = pytest_file.replace('.py', '')
     test(
       'pytest/@0@'.format(testname),
       pytest,
-      args: pytest_args + ['-k', testname],
+      args: [meson.current_source_dir()] + pytest_args + ['-k', testname],
       env: pytest_env,
       suite: ['pytest'],
       timeout: 120,
     )
   endforeach
+
+  if enable_installed_tests
+    install_data(
+      pytest_files,
+      '__init__.py',
+      'conftest.py',
+      'asan.suppression',
+      install_dir: installed_tests_dir / 'tests',
+    )
+    install_data(
+      template_files,
+      install_dir: installed_tests_dir / 'tests' / 'templates',
+    )
+
+    installed_env = {
+      'XDG_DESKTOP_PORTAL_PATH': libexecdir / 'xdg-desktop-portal',
+      'XDG_PERMISSION_STORE_PATH': libexecdir / 'xdg-permission-store',
+      'XDG_DOCUMENT_PORTAL_PATH': libexecdir / 'xdg-document-portal',
+      'XDP_VALIDATE_AUTO': '1',
+    }
+    env = ''
+    foreach key, value : installed_env
+      env += f'@key@=@value@ '
+    endforeach
+
+    foreach pytest_file : pytest_files
+      testname = pytest_file.replace('.py', '')
+
+      exec = [pytest.full_path(), installed_tests_dir / 'tests']
+      exec += pytest_args
+      exec += ['-o', 'cache_dir=/tmp/xdp-pytest-cache']
```

Comment:
Let's not introduce /tmp symlink attacks via hard-coded filenames in /tmp, please. ([protected_symlinks](https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-symlinks) mitigates this into only being a denial of service, but it's still best avoided.)

Can we use `-o cache_dir=.` or `-o cache_dir=/proc/self/cwd`, relying on the fact that ginsttest-runner runs each test with its current working directory set to a temporary directory that it automatically cleans up afterwards?

Or perhaps `-p no:cacheprovider`?

Reply:
> relying on the fact that ginsttest-runner runs each test with its current working directory set to a temporary directory

Reference: https://wiki.gnome.org/Initiatives(2f)GnomeGoals(2f)InstalledTests.html

Reply:
Testing https://github.com/smcv/xdg-desktop-portal/tree/wip/pytest-installed now

Reply:
Yes, that works. Please consider applying the last commit of that branch, either squashed into [build/tests: Install pytest based tests](https://github.com/flatpak/xdg-desktop-portal/pull/1525/commits/e190f5ce64848326cb2a5f439a3719fa94accff0) or separately.

Reply:
wtf. It works, and says `cachedir: .`, but I still get (non-fatal!) warnings about attempting to write into `/usr/libexec/installed-tests`. Sorry, I don't understand pytest well enough to diagnose this... but the tests pass so that's an excellent start.

Reply:
Trying now

Reply:
Yeah, this works nicely